projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,343 @@
+# e.g. [(discardedForSize, {"main": 3, "crash": 5}), (sendFailure, {"timeout" : 34})]
+def getFailuresStatPerFailureName(pings, failureNames):
+    def reduceFailure(failureName):
+        return pings.map(lambda p: p[failureName]).reduce(aggregateFailures)
+    
+    healthPingFailuresStat = map(reduceFailure, failuresNames)
+    return zip(failuresNames, healthPingFailuresStat)
+    
+```
+
+```python
+failuresNames = ["payload/pingDiscardedForSize", "payload/sendFailure"]
+```
+
+```python
+failuresStat = getFailuresStatPerFailureName(cachedData, failuresNames)
+```
+
+```python
+for fs in failuresStat:
+    plt.title(fs[0])
+    plt.bar(range(len(fs[1])), fs[1].values(), align='center')
+    plt.xticks(range(len(fs[1])), fs[1].keys())
+    plt.show()
+    
+```
+
+
+![png](images/output_8_0.png)
+
+
+
+
+![png](images/output_8_1.png)
+
+
+# Validate payload
+* for every field which should be non empthy check that
+* for payload/reason check that reason value among the expected ("immediate", "delayed", "shutdown")
+* check that both payload/sendFailure and payload/discardedForSize non empty
+* for both payload/sendFailure and payload/discardedForSize chech that count paramter has type int
+* for payload/sendFailure check that sendFailureType value among the expected ("eOK", "eRequest", "eUnreachable", "eChannelOpen", "eRedirect", "abort", "timeout")
+* for payload/discardedForSize check that we have only 10 records
+
+
+```python
+OK = ""
+MUST_NOT_BE_EMPTY = "must not be empty"
+    
+def validate(ping):
+    # validate os
+    os = ping["payload/os"]
+    if os == None:
+        return ("OS " + MUST_NOT_BE_EMPTY, ping)
+
+    name, version = os.items()
+    if name == None:
+        return ("OS name " + MUST_NOT_BE_EMPTY, ping)
+    if version == None:
+        return ("OS version " + MUST_NOT_BE_EMPTY, ping)
+    
+    # validate reason
+    reason = ping["payload/reason"]
+    if reason == None:
+        return ("Reason " + MUST_NOT_BE_EMPTY, ping)
+    
+    if not reason in ["immediate", "delayed", "shutdown"]:
+        return ("Reason must be equal to immediate, delayed or shutdown", ping)
+    
+    # doesn't contain failures
+    sendFailure = ping["payload/sendFailure"]
+    pingDiscardedForSize = ping["payload/pingDiscardedForSize"]
+    if sendFailure == None and pingDiscardedForSize == None:
+        return ("Ping must countain at least one of the failures", ping)
+
+    # validate sendFailure
+    if sendFailure != None and len(sendFailure) > 8:
+        return ("send Failure accept only 8 send failures", ping)
+    
+    supportedFailureTypes = ["eOK", "eRequest", "eUnreachable", "eChannelOpen", "eRedirect", "abort", "timeout"]
+    if sendFailure != None:
+        for key in sendFailure.keys():
+            if not key in supportedFailureTypes:
+                return (key + " type is not supported", ping)
+        for count in sendFailure.values():
+            if not isinstance(count, int):
```

Comment:
A distribution/histogram would be easy.

If you want a specific number, we could go from theoretical limits:
Ping sends are throttled to 10 per min, so we should not see more failures than 10 per min per Firefox instance.
Each health ping covers up to 1 hour, so we should probably not see much more than 60 sendFailures per ping.
Visualizing the distribution will allow us to verify that.